builtin/logical/pki/crl_util.go

```diff
@@ -26,6 +30,65 @@ type revocationInfo struct {
 	CertificateIssuer issuerID  `json:"issuer_id"`
 }
 
+// crlBuilder is gatekeeper for controlling various read/write operations to the storage of the CRL.
+// The extra complexity arises from secondary performance clusters seeing various writes to its storage
+// without the actual API calls. During the storage invalidation process, we do not have the required state
+// to actually rebuild the CRLs, so we need to schedule it in a deferred fashion. This allows either
+// read or write calls to perform the operation if required, or have the flag reset upon a write operation
+type crlBuilder struct {
+	m            sync.Mutex
+	forceRebuild uint32
+}
+
+const (
+	_ignoreForceFlag  = true
+	_enforceForceFlag = false
+)
+
+// rebuildIfForced is to be called by readers or periodic functions that might need to trigger
+// a refresh of the CRL before the read occurs.
+func (cb *crlBuilder) rebuildIfForced(ctx context.Context, b *backend, request *logical.Request) error {
+	if atomic.LoadUint32(&cb.forceRebuild) == 1 {
```

Comment:
`rebuildIfForced` is called from a number of contexts/places where, if we aren't _asked_ to rebuild the CRL, we shouldn't (e.g., the period function and/or CRL fetching has a hook to this call -- because the invalidate function doesn't have access to the relevant data for managed keys). 

We'd need to pair a `RWLock` with a `bool`, which is equivalent to an `atomic` int-ish, I think?

Reply:
Correct, what @cipherboy said. We are kinda trying to recreate a sync.Once with resettable behaviour along with the ability. So both a RWLock would not be good enough as the `invalidate` call mainly doesn't have the context (missing storage and more importantly mountpoint) to rebuild the CRL. A sync.Once wouldn't be good enough as we need to redo this operation/reset it on subsequent `invalidate` calls in a safe manner.

Reply:
I think from the earlier call, a `sync.Once` can be a pointer to a `sync.Once` that allows it to be "resettable", though now you need an atomic "write pointer" I think?

Reply:
Correct, but our use case isn't a direct match either for a sync.Once as the initial state, say for a single node, would be to never execute under a read api call occurred. That's another reason I did not use an actual sync.Once as the initialization code would have had to execute something arbitrary and then allow it to be reset.

Reply:
Good enough for me.